docs/storage-config.md

```diff
@@ -0,0 +1,267 @@
+QEMU process is living in the libvirtd POD namespaces, any local storage has to
+be made available at the time the libvirtd POD is started. Volumes that are made
+available to the VM POD are not directly accessible from the libvirtd POD. There
+may be some tricks that can be used to get access to VM POD storage, from the
+libvirtd POD, but these would likely have to rely on internal impl details of
+k8s which are not guaranteed / supported interfaces, so not desirable to do.
+
+At this time there is no way to make host block devices directly accessible to
+PODs. If this is enabled in the future, it will share the same problem as local
+file storage, in that the libvirtd POD cannot access block devices associated
+with the VM POD, so any block devices used would have to be devices that are
+visible to the libvirtd POD when it is initially launched.
+
+### Direct network block devices
+
+Given the complications with local file/block storage, the simplest way to
+configure a virtual disk is using a direct network block device. In the VM spec
+this can be configured two ways.
+
+#### Direct storage reference
+
+The storage server can be referenced directly, for example, as an iSCSI server
+address:
+
+```yaml
+    disks:
+      - source:
+          iscsi:
+            host:
+              name: iscsi-demo-target
+              port: "3260"
+            portal: iqn.2017-01.io.kubevirt:sn.42
+	    lun: 2
+```
+
+This would also work for RBD, NBD, GlusterFS, and HTTP(s) protocols.
+
+#### Indirect storage reference
+
+The k8s PersistentVolumeClaim (PVC) resource is an abstract concept that allows
+a tenant user to claim a PersistentVolume (PV). There are many different
+backends for PVs, some are filesystem based and some are network block device
+based. Using QEMU's direct network backend drivers, it is possible to directly
+access the storage for any PV that is backend by a network block device (in
+particular the iSCSI and RBD k8s backends).
+
+```yaml
+    disks:
+      - source:
+          pvc:
+	    claimname: some-pvc-name
+```
+
+When launching QEMU the PVC resource is accessed to get the corresponding
+PV resource. The configuration for the ISCSI / RBD backend is then extracted
+to configure the QEMU storage backend
+
+
+### Dynamic storage reference
+
+Rather than having the storage volume pre-created upfront, it may be useful to
+have a way to dynamically create storage backends. To achieve this, it is
+possible to define a storage source which defines a container. The virt-handler
+would create an instance of this container, which would be required to expose a
+HTTP service with a well-known URI eg https://$POD-IP/volume
+
+Making a PUT request to this URI would trigger creation of the volume, and a
+a DELETE request would naturally delete it. The response to the PUT request
+would be a YAML document following the 'iscsi' or 'pvc' schema for the VM
+spec disk source. eg
```

Comment:
Definitely not virt-handler. virt-controller could help. There are differnet ways, but none of them require such a REST api on the virt-handler side, which like stated further down in another comment, moves responsibilities to the node, which are not node-related.

So there are different ways, how such flows can be enabled in a more k8s like approach:

1) Dynamic provisioners (with storage class)
2) full reedom, by doing your completely own storage with your own services and pods
3) Integrating e.g. a provision container into the VM spec, which will force virt-controller to create and provision the appropriate PVCs before the VM is even scheduled. Also note that a potential required delete of such a storage, after a VM stopped, would be done by a controller, not virt-handler.

Possible example:

```yaml
apiVersion: kubevirt.io/v1alpha1
kind: VM
metadata:
  name: testvm
spec:
  domain:
    ...
  storage:
    recycleStrategy: deleteOnVMStop
    template:
      ... my provisioner pod template ...
```
However, there might be reason, why we would still want to completely keep the storage lifetime and the VM spec separated. In this case the next option would also be interesting.

4) Create something like a Disk TPR, which contains a provisioner pod template:

```yaml
apiVersion: kubevirt.io/v1alpha1
kind: ProvisionedDisk
metadata:
  name: testvm
spec:
  pvc:
    template:
    ... my pvc template ...
  provisioner:
    template:
      ... my provisioner pod template ...
status:
  phase: Bound
```
Bound in this case could mean for virt-handler, PVC is ready to go and provisioned, so resolve it and use it.


However, the main point is, that whenever we want to give the user such tools, we have to think about letting controllers do the cluster wide stuff and let the controller object status represent, when the resource is ready. virt-handler can then just consume it. Other things can be solved on the node. They should be done by virt-handler directly. A special rest api which virt-handler would query, does absolutely not make sense in any way. 

Something which in my eyes is a very good decision tree, on how to integrate with k8s is: First think about what the user (operator) want's to solve and what the steps are, then think about the scope (cluster vs node). Then, if it is a cluster scope, think about a controller which can do the steps for the operator.

Finally, there might still be the need for management applications, which want to use kubevirt, to provide something like an  upload rest-api to the user. All they would have to do, is to write their upload application, deploy it to k8s, configure a controller like outlined in (3) and (4) with the right pod template (to copy uploaded images to the right spot in a PVC, or transform the PVC into the image), an they are done.